### PR TITLE
Adds elasticsearch to the user and channel autocompletion functions

### PR DESCRIPTION
**Pull request by mgdelacroix**
#### Summary
This PR creates templates for users and channels in Elasticsearch, and if the indexing is enabled by config, indexes both entities and provides autocompletion functions through the Elasticsearch interface

#### Ticket Link
This ticket covers: [MM-13629](https://mattermost.atlassian.net/browse/MM-13629), [MM-13630](https://mattermost.atlassian.net/browse/MM-13630), [MM-13631](https://mattermost.atlassian.net/browse/MM-13631), [MM-13632](https://mattermost.atlassian.net/browse/MM-13632), [MM-13633](https://mattermost.atlassian.net/browse/MM-13633)

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [x] [Has enterprise changes](https://github.com/mattermost/enterprise/pull/399)

### CODE REVIEW

File: app/channel.go [hunk 1]
```diff
@@ -223,6 +241,24 @@ func (a *App) CreateChannel(channel *model.Channel, addMember bool) (*model.Chan
 		})
 	}
 
+	esInterface := a.Elasticsearch
+	if esInterface != nil && *a.Config().ElasticsearchSettings.EnableIndexing {
+		if sc.Type == "O" {
```
Comment (grundleborg) [inline app/channel.go]: Why this conditional check?

File: app/channel.go [hunk 2]
```diff
@@ -431,6 +489,15 @@ func (a *App) UpdateChannel(channel *model.Channel) (*model.Channel, *model.AppE
 	messageWs.Add("channel", channel.ToJson())
 	a.Publish(messageWs)
 
+	esInterface := a.Elasticsearch
+	if esInterface != nil && *a.Config().ElasticsearchSettings.EnableIndexing && channel.Type == "O" {
```
Comment (grundleborg) [inline app/channel.go]: Same question here.

File: app/channel.go [hunk 3]
```diff
@@ -1585,6 +1679,31 @@ func (a *App) UpdateChannelLastViewedAt(channelIds []string, userId string) *mod
...
+	license := a.License()
+	if esInterface != nil && *a.Config().ElasticsearchSettings.EnableAutocomplete && license != nil && *license.Features.Elasticsearch {
+		channelIds, err := a.Elasticsearch.SearchChannels(teamId, term)
+		if err != nil {
+			return nil, err
+		}
+
+		channelList := model.ChannelList{}
+		if len(channelIds) > 0 {
+			cresult := <-a.Srv.Store.Channel().GetChannelsByIds(channelIds)
```
Comment (grundleborg) [inline app/channel.go]: Is there a "worst case" maximum length for `channelIds` enforced somewhere before this query?

File: app/channel.go (hunk 1 above)
Comment (mgdelacroix) [reply app/channel.go]: My understanding is that autocompletion only happens for public or open channels, so this check exists to only index the open ones.

File: app/channel.go (hunk 3 above)
Comment (mgdelacroix) [reply app/channel.go]: There is not, and it should be. The non-elasticsearch behavior for autocompletion is limiting the results at `50`, so we can replicate this directly on the elastic queries using `Size`

File: app/channel.go (hunk 1 above)
Comment (grundleborg) [reply app/channel.go]: Oh yeah, seems you are right (although I'm quite surprised by that!).

File: app/channel.go (hunk 3 above)
Comment (grundleborg) [reply app/channel.go]: :+1:

**Comment by cpanato** [comment]
test comment, please ignore

**Pull request merged**